File: Тренировки к МОШ/Marathon/2019/F/main.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <bitset>
using namespace std;

map<int, bool> used;
string coms;
vector<int> order;
map<int, int> tin, tout;
int timer = 0;

void rec(int v, int idx) {
    if (used.find(v) == used.end()) {
        used[v] = true;
    }
    tin[v] = timer++;
    order.push_back(v);
    if (coms[idx] == 'D') {
        if (used.find(v * 2) == used.end()) {
            rec(v * 2, idx + 1);
        } else {
            rec(v * 2 + 1, idx + 1);
        }
    } else {
        if (v % 2 == 0) {
            rec(v + 1, idx + 1);
        } else {
            while (v % 2 != 0) {
                v /= 2;
            }
            rec(v, idx + 1);
        }
    }
    tout[v] = timer++;
}
void solve() {
    used.clear();
    order.clear();
    tin.clear(), tout.clear();
    timer = 0;
    cin >> coms;
    rec(1, 0);
    vector<long long> ans;
    for (auto v : order) {
        if (tin[v] = tout[v] - 1) {
            cout << get(v) << "\n";
        }
    }


}
int main() {
    int q;
    cin >> q;
    while (q--) {
        solve();
    }
}

File: Тренировки к МОШ/Marathon/2019/F/main.py
q = int(input())

used = {}
coms = ""
ans = []

def get(x):
    res = ""
    while x != 1:
        if x % 2 == 0:
            res += "0"
        else:
            res += "1"
        x //= 2
    return res[::-1]
def rec(v, idx):
    global used, coms, ans
    if v not in used:
        used[v] = True
    if idx == len(coms):
        ans.append(v)
        return
    if coms[idx] == 'D':
        if v * 2 not in used:
            rec(v * 2, idx + 1)
        else:
            rec(v * 2 + 1, idx + 1)
    else:
        ans.append(v)
        if v % 2 == 0:
            rec(v + 1, idx + 1)
        else:
            while v % 2 != 0:
                v //= 2
            rec(v + 1, idx + 1)
for _ in range(q):
    used = {}
    coms = ""
    coms = input()
    ans = []
    rec(1, 0)
    #print(ans)
    print(len(ans))
    for x in ans:
        print(get(x))
    print()
    